core/metacling/src/TClingCallbacks.cxx

```diff
@@ -100,7 +100,9 @@ void TClingCallbacks::InclusionDirective(clang::SourceLocation sLoc/*HashLoc*/,
    //    or TH1F in presence of TH1F.h.
    // Strategy 2) is tried only if 1) fails.
 
-   if (!IsAutoloadingEnabled() || fIsAutoloadingRecursively || !FileName.endswith(".h")) return;
+   bool isHeaderFile = FileName.endswith(".h") || FileName.endswith(".hxx")
+                        || FileName.endswith(".hpp");
+   if (!IsAutoloadingEnabled() || fIsAutoloadingRecursively || !isHeaderFile) return;
 
    std::string localString(FileName.str());
```

Comment:
It looks like the changes above would only make sense if we actually strip the extension from FileName here. We should look for `TObject` if FileName is `TObject.h`.

Reply:
Why?

Reply:
We are in case 2) described in the documentation. The lines below seem to be performing a Sema Lookup and I do not see why we should be doing a lookup based on LookupResult containing `TObject.h`.

Reply:
I see.  You are partially right.  The first  argument to  tryAutoParseInternal still needs to be the header filename.